File: iq/list/singly/flatten/do_test.go
// Copyright 2025 Samvel Khalatyan. All rights reserved.

package flatten_test

import (
	"testing"

	"github.com/google/go-cmp/cmp"
	"github.com/skhal/lab/iq/list/singly/flatten"
)

var toSlice = cmp.Transformer("ToSlice", func(n *flatten.Node) []int {
	return n.Slice()
})

func TestDo_empty(t *testing.T) {
	tree := flatten.NewTree()
	want := flatten.NewList()

	got := flatten.Do(tree)

	if diff := cmp.Diff(want, got, toSlice); diff != "" {
		t.Errorf("flatten.Do() mismatch (-want, +got):\n%s", diff)
		t.Logf("tree:\n%v", tree)
	}
}

func TestDo_flatOneItem(t *testing.T) {
	tree := flatten.NewTree(1)
	want := flatten.NewList(1)

	got := flatten.Do(tree)

	if diff := cmp.Diff(want, got, toSlice); diff != "" {
		t.Errorf("flatten.Do() mismatch (-want, +got):\n%s", diff)
		t.Logf("tree:\n%v", tree)
	}
}

func TestDo_flatTwoItems(t *testing.T) {
	tree := flatten.NewTree(1, 2)
	want := flatten.NewList(1, 2)

	got := flatten.Do(tree)

	if diff := cmp.Diff(want, got, toSlice); diff != "" {
		t.Errorf("flatten.Do() mismatch (-want, +got):\n%s", diff)
		t.Logf("tree:\n%v", tree)
	}
}

func TestDo_flatThreeItems(t *testing.T) {
	tree := flatten.NewTree(1, 2, 3)
	want := flatten.NewList(1, 2, 3)

	got := flatten.Do(tree)

	if diff := cmp.Diff(want, got, toSlice); diff != "" {
		t.Errorf("flatten.Do() mismatch (-want, +got):\n%s", diff)
		t.Logf("tree:\n%v", tree)
	}
}

func TestDo_l1OneL2One(t *testing.T) {
	tree := flatten.NewTree(1)
	tree.Get(1).SetChild(flatten.NewTree(2))
	want := flatten.NewList(1, 2)

	got := flatten.Do(tree)

	if diff := cmp.Diff(want, got, toSlice); diff != "" {
		t.Errorf("flatten.Do() mismatch (-want, +got):\n%s", diff)
		t.Logf("tree:\n%v", tree)
	}
}

func TestDo_l1TwoL2OneOnFirstL1(t *testing.T) {
	tree := flatten.NewTree(1, 2)
	tree.Get(1).SetChild(flatten.NewTree(3))
	want := flatten.NewList(1, 2, 3)

	got := flatten.Do(tree)

	if diff := cmp.Diff(want, got, toSlice); diff != "" {
		t.Errorf("flatten.Do() mismatch (-want, +got):\n%s", diff)
		t.Logf("tree:\n%v", tree)
	}
}

func TestDo_l1TwoL2OneOnSecondL1(t *testing.T) {
	tree := flatten.NewTree(1, 2)
	tree.Get(2).SetChild(flatten.NewTree(3))
	want := flatten.NewList(1, 2, 3)

	got := flatten.Do(tree)

	if diff := cmp.Diff(want, got, toSlice); diff != "" {
		t.Errorf("flatten.Do() mismatch (-want, +got):\n%s", diff)
		t.Logf("tree:\n%v", tree)
	}
}

func TestDo_l1TwoL2TwoOnFirstL1(t *testing.T) {
	tree := flatten.NewTree(1, 2)
	tree.Get(1).SetChild(flatten.NewTree(3, 4))
	want := flatten.NewList(1, 2, 3, 4)

	got := flatten.Do(tree)

	if diff := cmp.Diff(want, got, toSlice); diff != "" {
		t.Errorf("flatten.Do() mismatch (-want, +got):\n%s", diff)
		t.Logf("tree:\n%v", tree)
	}
}

func TestDo_l1TwoL2TwoOnSecondL1(t *testing.T) {
	tree := flatten.NewTree(1, 2)
	tree.Get(2).SetChild(flatten.NewTree(3, 4))
	want := flatten.NewList(1, 2, 3, 4)

	got := flatten.Do(tree)

	if diff := cmp.Diff(want, got, toSlice); diff != "" {
		t.Errorf("flatten.Do() mismatch (-want, +got):\n%s", diff)
		t.Logf("tree:\n%v", tree)
	}
}

func TestDo_l1TwoL2OnePerL1Item(t *testing.T) {
	tree := flatten.NewTree(1, 2)
	tree.Get(1).SetChild(flatten.NewTree(3))
	tree.Get(2).SetChild(flatten.NewTree(4))
	want := flatten.NewList(1, 2, 3, 4)

	got := flatten.Do(tree)

	if diff := cmp.Diff(want, got, toSlice); diff != "" {
		t.Errorf("flatten.Do() mismatch (-want, +got):\n%s", diff)
		t.Logf("tree:\n%v", tree)
	}
}

File: iq/list/singly/flatten/do_test.cc
// Copyright 2025 Samvel Khalatyan. All rights reserved.
//
// clang-format off-next-line
//go:build ignore

#include "iq/list/singly/flatten/do.h"

#include <cctype>
#include <initializer_list>
#include <memory>
#include <vector>

#include "gmock/gmock.h"
#include "gtest/gtest.h"

namespace iq::list::singly::flatten {
namespace {

using ::testing::ContainerEq;
using ::testing::IsEmpty;

std::shared_ptr<TreeNode> makeTree(std::initializer_list<int> nn) {
  std::shared_ptr<TreeNode> head;
  std::shared_ptr<TreeNode> tail;
  for (int n : nn) {
    auto node = std::make_shared<TreeNode>(n);
    if (!head) {
      head = node;
      tail = node;
      continue;
    }
    tail->next = node;
    tail = node;
  }
  return head;
}

std::shared_ptr<TreeNode> getNode(std::shared_ptr<TreeNode> node, int val) {
  for (; node; node = node->next) {
    if (node->val == val) {
      return node;
    }
  }
  return nullptr;
}

std::vector<int> toVector(std::shared_ptr<Node> list) {
  std::vector<int> nn;
  for (; list; list = list->next) {
    nn.emplace_back(list->val);
  }
  return nn;
}

MATCHER_P(ListEq, want, "") {
  const std::vector<int> got = toVector(arg);
  EXPECT_THAT(got, ContainerEq(want));
  return true;
}

MATCHER(IsEmptyList, "") {
  const std::vector<int> got = toVector(arg);
  EXPECT_THAT(got, IsEmpty());
  return true;
}

TEST(DoTest, Empty) {
  auto tree = makeTree({});

  std::shared_ptr<Node> got = Do(tree);

  EXPECT_THAT(got, IsEmptyList());
}

TEST(DoTest, FlatOneItem) {
  auto tree = makeTree({1});

  std::shared_ptr<Node> got = Do(tree);

  EXPECT_THAT(got, ListEq(std::vector<int>{1}));
}

TEST(DoTest, FlatTwoItems) {
  auto tree = makeTree({1, 2});

  std::shared_ptr<Node> got = Do(tree);

  EXPECT_THAT(got, ListEq(std::vector<int>{1, 2}));
}

TEST(DoTest, FlatThreeItems) {
  auto tree = makeTree({1, 2, 3});

  std::shared_ptr<Node> got = Do(tree);

  EXPECT_THAT(got, ListEq(std::vector<int>{1, 2, 3}));
}

TEST(DoTest, Level1OneLevel2One) {
  auto tree = makeTree({1});
  getNode(tree, 1)->child = makeTree({2});

  std::shared_ptr<Node> got = Do(tree);

  EXPECT_THAT(got, ListEq(std::vector<int>{1, 2}));
}

TEST(DoTest, Level1TwoLeve2OneOnFirst) {
  auto tree = makeTree({1, 2});
  getNode(tree, 1)->child = makeTree({3});

  std::shared_ptr<Node> got = Do(tree);

  EXPECT_THAT(got, ListEq(std::vector<int>{1, 2, 3}));
}

TEST(DoTest, Level1TwoLeve2OneOnSecond) {
  auto tree = makeTree({1, 2});
  getNode(tree, 2)->child = makeTree({3});

  std::shared_ptr<Node> got = Do(tree);

  EXPECT_THAT(got, ListEq(std::vector<int>{1, 2, 3}));
}

TEST(DoTest, Level1TwoLeve2TwoOnFirst) {
  auto tree = makeTree({1, 2});
  getNode(tree, 1)->child = makeTree({3, 4});

  std::shared_ptr<Node> got = Do(tree);

  EXPECT_THAT(got, ListEq(std::vector<int>{1, 2, 3, 4}));
}

TEST(DoTest, Level1TwoLeve2TwoOnSecond) {
  auto tree = makeTree({1, 2});
  getNode(tree, 2)->child = makeTree({3, 4});

  std::shared_ptr<Node> got = Do(tree);

  EXPECT_THAT(got, ListEq(std::vector<int>{1, 2, 3, 4}));
}

TEST(DoTest, Level1TwoLeve2TwoOnEach) {
  auto tree = makeTree({1, 2});
  getNode(tree, 1)->child = makeTree({3});
  getNode(tree, 2)->child = makeTree({4});

  std::shared_ptr<Node> got = Do(tree);

  EXPECT_THAT(got, ListEq(std::vector<int>{1, 2, 3, 4}));
}

}  // namespace
}  // namespace iq::list::singly::flatten
